clio-server/src/main/scala/org/broadinstitute/clio/server/service/IndexService.scala

```diff
@@ -67,4 +99,58 @@ abstract class IndexService[CI <: ClioIndex](
       elasticsearchIndex
     )
   }
+
+  def validateUpsert(
+    existingMetadata: Json,
+    newMetadata: Json
+  ): Either[Exception, Json] = {
+    val differences = JsonDiff.diff(existingMetadata, newMetadata, remember = true)
+
+    if (differences.ops.nonEmpty) {
+      Right(newMetadata)
+    } else {
+      val diffs: String = differences.ops.map { op =>
+        op.toString
+      }.mkString("\n")
+      Left(
+        UpsertValidationException(
+          s"""Adding this document will overwrite the following existing metadata:
+               | $diffs. Use 'force=true' to overwrite the existing data.""".stripMargin
+        )
+      )
+    }
+  }
+
+  def queryMetadataForKey(
+    indexKey: clioIndex.KeyType
+  ): Source[Json, NotUsed] = {
+    val keyJson = indexKey.asJsonObject
+    val fieldsToDrop = keyJson.keys.toSeq
+
+    searchService
+      .simpleStringQueryMetadata(indexKey.asJson.toString())(
```

Comment:
I'm not sure `toString()` will do what we want. I think `ModelAutoDerivation.defaultPrinter.noSpaces(indexKey.asJson)` will do the trick.

Reply:
hmm weird.. it seemed to work fine when i was testing it

Reply:
TIL! https://github.com/circe/circe/blob/master/modules/core/shared/src/main/scala/io/circe/Json.scala#L154